projects/packages/backup/src/class-rest-controller.php

```diff
@@ -138,4 +255,277 @@ public static function delete_backup_helper_script( $request ) {
 			)
 		);
 	}
+
+	/**
+	 * Fetch a backup of a database object, along with all of its metadata.
+	 *
+	 * @access public
+	 * @static
+	 *
+	 * @param WP_REST_Request $request The request sent to the WP REST API.
+	 * @return array
+	 */
+	public static function fetch_database_object_backup( $request ) {
+		global $wpdb;
+
+		// Disable Sync as this is a read-only operation and triggered by sync activity.
+		Sync_Actions::mark_sync_read_only();
+
+		$allowed_object_types = self::get_allowed_object_types();
+		// Safe to do this as we have already validated the object_type key exists in self::get_allowed_object_types().
+		$object_type = $allowed_object_types[ $request->get_param( 'object_type' ) ];
+		$object_id   = $request->get_param( 'object_id' );
+		$table       = $wpdb->prefix . $object_type['table'];
+		$id_field    = $object_type['id_field'];
+
+		// Fetch the requested object.
+		$object = $wpdb->get_row(
+			$wpdb->prepare(
+				'SELECT * FROM `' . $table . '` WHERE `' . $id_field . '` = %d', // phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
+				$object_id
+			)
+		);
+
+		if ( empty( $object ) ) {
+			return new WP_Error( 'object_not_found', __( 'Object not found', 'jetpack' ), array( 'status' => 404 ) );
+		}
+
+		$result = array( 'object' => $object );
+
+		// Fetch associated metadata (if this object type has any).
+		if ( ! empty( $object_type['meta_type'] ) ) {
+			$result['meta'] = get_metadata( $object_type['meta_type'], $object_id );
+		}
+
+		// If there is a child linked table (eg: woocommerce_tax_rate_locations), fetch linked records.
+		if ( ! empty( $object_type['child_table'] ) ) {
+			$child_table    = $wpdb->prefix . $object_type['child_table'];
+			$child_id_field = $object_type['child_id_field'];
+
+			$result['children'] = $wpdb->get_results(
+				$wpdb->prepare(
+					'SELECT * FROM `' . $child_table . '` where `' . $child_id_field . '` = %d', // phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
+					$object_id
+				)
+			);
+		}
+
+		return $result;
+	}
+
+	/**
+	 * Fetch a backup of an option.
+	 *
+	 * @access public
+	 * @static
+	 *
+	 * @param WP_REST_Request $request The request sent to the WP REST API.
+	 * @return array
+	 */
+	public static function fetch_options_backup( $request ) {
+		// Disable Sync as this is a read-only operation and triggered by sync activity.
+		Sync_Actions::mark_sync_read_only();
+
+		$option_name = $request->get_param( 'name' );
+		if ( is_array( $option_name ) ) {
+			$option_names = $option_name;
+		} else {
+			$option_names = array( $option_name );
+		}
+
+		$options = array_map( 'self::get_option_row', $option_names );
+		return array( 'options' => $options );
```

Comment:
If we query a non-existing option, we still get one empty option element:
```
"body":{"options":[null]}
```

<img width="157" alt="Screen Shot 2021-06-29 at 12 03 37 PM" src="https://user-images.githubusercontent.com/1341249/123769536-11e6c400-d8d2-11eb-91f3-62f863db6b59.png">

Should we validate that options in fact exist before returning them?

Reply:
Nice catch, Sergey! I agree with you, but I'm a bit hesitant to change that response thinking about backwards compatibility with previous JP versions that will still be using the old endpoint.

It's possible that consumers expect the above response when querying a non-existing option so I think it should be identical between our new REST endpoint and the existing one.

Reply:
Makes sense, let's keep it as is.
I hate that when we find a bug and can't fix it because it would broke the backward compatibility 🙂